synfig-studio/src/gui/dialogs/canvasresize.cpp

```diff
@@ -0,0 +1,548 @@
+
+		is_canvas_checked ? set_canvas_flags(true) : set_image_flags(true);
+	} else
+		is_canvas_checked ? set_canvas_flags(false) : set_image_flags(false);
+}
+
+void CanvasResize::on_spinbutton_updated(Gtk::SpinButton *widget)
+{
+	// Prevent widget from adding default value 1 if user fully overwrites w or h
+	if (widget->get_text_length() == 0) return;
+
+	if (is_canvas_centered && !is_canvas_ready)
+		is_canvas_ready = true;
+
+	width ->update();
+	height->update();
+}
+
+void CanvasResize::refresh_title()
+{
+	set_title(_("Resize - ") + canvas_interface->get_canvas()->get_name());
+}
+
+void CanvasResize::refresh_content()
+{
+	// Canvas button grid
+	int  index;
+	auto coordinates = rend_desc.get_focus();
+	auto x           = coordinates[0];
+	auto y           = coordinates[1];
+
+	if (x < 0 && y > 0)
+		index = 0;
+	else if (x == 0 && y == rend_desc.get_tl()[1])
+		index = 1;
+	else if (x > 0 && y > 0)
+		index = 2;
+	else if (x == rend_desc.get_tl()[0] && y == 0)
+		index = 3;
+	else if (x == rend_desc.get_br()[0] && y == 0)
+		index = 5;
+	else if (x < 0 && y < 0)
+		index = 6;
+	else if (x == 0 && y == rend_desc.get_br()[1])
+		index = 7;
+	else if (x > 0 && y < 0)
+		index = 8;
+	else
+		index = 4;
+
+	reset_canvas_buttons();
+
+	canvas_buttons[index]->get_image()->set_name("center");
+
+	canvas_center = canvas_buttons[index];
+
+	is_canvas_checked ? set_canvas_button_images() : set_canvas_disabled_button_image();
+
+	refresh_wh_toggle_widgets();
+}
+
+void CanvasResize::refresh_wh_toggle_widgets()
+{
+	UpdateLock lock(update_lock);
+
+	width ->set_value(rend_desc.get_w());
+	height->set_value(rend_desc.get_h());
+	toggle_ratio_wh->set_active(rend_desc.get_flags() & RendDesc::LINK_IM_ASPECT);
+}
+
+void CanvasResize::reset_canvas_buttons()
+{
+	// Hide all our button images and reset center point
+	for (auto button : canvas_buttons) {
+		auto image = static_cast<Gtk::Image *>(button->get_image());
+		if (image->get_visible())       image->hide();
+		if (!image->get_name().empty()) image->unset_name();
+	}
+}
+
+void CanvasResize::set_canvas_center_point()
+{
+	if (canvas_center == canvas_buttons[0])
+		rend_desc.set_focus(rend_desc.get_tl());
+	else if (canvas_center == canvas_buttons[1])
+		rend_desc.set_focus(Point(0, rend_desc.get_tl()[1]));
+	else if (canvas_center == canvas_buttons[2])
+		rend_desc.set_focus(Point(rend_desc.get_br()[0], rend_desc.get_tl()[1]));
+	else if (canvas_center == canvas_buttons[3])
+		rend_desc.set_focus(Point(rend_desc.get_tl()[0], 0));
+	else if (canvas_center == canvas_buttons[5])
+		rend_desc.set_focus(Point(rend_desc.get_br()[0], 0));
+	else if (canvas_center == canvas_buttons[6])
+		rend_desc.set_focus(Point(rend_desc.get_tl()[0], rend_desc.get_br()[1]));
+	else if (canvas_center == canvas_buttons[7])
+		rend_desc.set_focus(Point(0, rend_desc.get_br()[1]));
+	else if (canvas_center == canvas_buttons[8])
+		rend_desc.set_focus(rend_desc.get_br());
+	else
+		rend_desc.set_focus(Point(0, 0));
+}
```

Comment:
I particularly am not sure about changing the RenderDesc focus point. Although almost nobody uses it…

Reply:
> Although almost nobody uses it…

Now everybody will use it but abstracted :)

If someone wants to change back to focus(0,0) simple; just click the center button and apply or s/he could do it from Canvas Properties too.

Earlier I manipulated tl and br directly, that works too but shrinking resolution was a bit hackish.